✨ feat: printing window titles

cseriildi opened the pull request:
- Titles are centered and bold.
- The board window shows the selected game mode.
- Scrolling still works, but the resize PR might break it

Edit:
#5 should be fine after this
added title wrap if the title is longer than the width of the window

itislu committed:
fix: Stop program when CTRL+D also in board building

itislu committed:
refactor: Rearrange jobs of `read_*` functions

Deciding to read from stdin or from a file should not be in the `read_file` function.
Same for resetting stdin to make it interactive.

Commit:
✨ feat: option selection with mouse

Commit:
🔧 chore: uploading map generator script

Commit:
🐛 fix: register mouse click immediately

https://community.unix.com/t/ncurses-how-shorten-mouse-click-time-response/384065/4

Commit:
✨ feat: printing window titles

itislu committed:
fix: Accept EOF in non-interactive mode anywhere to end board creation

The issue was when reading from stdin, we didn't know if it's interactive and want double newline to finish the board creation, or if it's coming from a file redirect and we accept EOF anywhere.

I added a function which should reliably detect if `STDIN_FILENO` is interactive or not, with the limited functions we are allowed to use.

With this I was also able to get rid of the global `g_stdin` variable.

itislu committed:
fix: Add missed `malloc()` failure detection when reading file

itislu committed:
fix: Fix missing empty line before board printout when read from file

Bug was introduced in commit 54a236356f05b90ead6d0a7d1e55f3a53d80c1d4.

itislu committed:
docs: Make comment about open-flags more clear

itislu committed:
fix(mandatory): Don't print game mode when user exits in selection

itislu committed:
improve: Make title printout border spacing consistent

itislu committed:
build: Add ncurses dependency to `bear` Dockerfile

itislu commented:
- If you think https://github.com/itislu/42-Rush-AlCu/pull/5 might brake it, I would say let's wait until that one is merged.
If you agree, you could mark this PR as a draft, so it doesn't get merged by accident.

Commit:
🐛 fix: title wrapping

Commit:
Merge branch 'main' into feat-printing-centered-title

Commit:
🐛 fix: typo

itislu commented on bonus/src/utils.c:
```diff
@@ -33,7 +33,28 @@ void print_title(t_win *win, char *title)
+			i = win->size.x - 2;
+		}
+		ft_strlcpy(tmp, title + start, i);
+		tmp[i] = '\0';
```
Just to double check: 
1. `(ft_)strlcpy()` nul-terminates the string, so manual nul-termination should not be necessary. If it is, it would be a hint that sth is wrong.
2. The size argument to `(ft_)strlcpy()` should include space for the nul-terminator. Is it included here?

![image](https://github.com/user-attachments/assets/b49b6933-cb7b-48e2-ba4b-b9a59f1501d8)

cseriildi replied:
valid point. Fixed it

itislu commented:
Just a side note: It looks like you merged main into this branch locally with `git merge main`.
By default, git uses fast-forward merge if it can. This means it will not create a merge commit. Bc of that, a bunch of commits that come from the main branch show up in this PR now.

This can be avoided by doing `git merge --no-ff main`. This is also what GitHub does when updating a PR branch by merge instead of rebase.

Commit:
🐛 fix: use ft_strlcpy properly

Commit:
🔧 chore: merge with win_resize

itislu commented on bonus/src/update_windows.c:
```diff
@@ -70,18 +68,22 @@ void update_board(t_win *n_board, t_board *board)
-	default:
+		default:
 		print_title(n_board, NO_GAME_MODE);
 	}
```
Fantastic short article if it's good to indent the `case` labels or not: https://www.learncpp.com/cpp-tutorial/switch-statement-basics/#:~:text=Labels%20are%20conventionally%20not%20indented

cseriildi replied:
A refactor is much needed anyway but good to know. It's fixed

Commit:
🎨 style: implementing switch case indentation convention

https://www.learncpp.com/cpp-tutorial/switch-statement-basics/#:~:text=Labels%20are%20conventionally%20not%20indented

cseriildi commented:
I've found some issues with the printing

Commit:
🐛 fix: avoid printing on the box

happened when game over in history window and when the board is the same size as the window

itislu commented on bonus/inc/visuals.h:
```diff
@@ -8,6 +8,14 @@
 #define MIN_TERMINAL_HEIGTH             9
 #define MIN_TERMINAL_WIDTH              MIN_BOARD_WIDTH + MIN_INFO_PANEL_WIDTH
 
+#define NO_GAME_MODE "Game mode is not selected yet"
```
Can we remove the `is` from this message? Fits better to the others.

And it would also fix this issue (increasing the minimum width by 1 is also possible):

https://github.com/user-attachments/assets/df6c2124-aed5-4478-80b7-aed7367a72ac

```suggestion
#define NO_GAME_MODE "Game mode not selected yet"
```

cseriildi replied:
Sure we can, but I’ll also want to see when did I brake that.

itislu replied:
I think this is a logic error, in how we decide how many lines to print, that was always there. Just nothing was able to ever show it.

cseriildi replied:
I removed the is, but also fixed the multiline thing

Commit:
🐛 fix: multiline title

scrolling after resizing with multiline titles

itislu commented:
@dltslt Bc a lot of the changes touch code that you wrote, especially bed2af3, please also take a look.

itislu commented:
I am in favor of self documenting code, writing comments only when necessary.
So I like the reduction of comments.

I read this this is generally good practice bc comments are very easy to become out-of-date bc the compiler doesn't care about them. And if a comment does not match the code anymore, it does more harm than good.

I believe comments should also describe on a higher level **what** something does, not **how** it does it.
And what the expected inputs and outputs of a function or a long block of code are.

Unless in very specific cases, like performance optimization, if a piece of code requires line by line comments to be readable, the code should probably be adjusted.

But tbh I am terrible at documenting code well, so take everything I write with a grain of salt.
Just wanted to share my point of view.

itislu committed:
Merge branch 'main' into feat-printing-centered-title

Commit:
Merge branch 'main' into feat-printing-centered-title

Commit:
🎨 style: making macro of the history window title to make it more uniform
